File: src/store.js
import fs from 'node:fs';
import path from 'node:path';
import Papa from 'papaparse';

const CSV_PATH = path.join(process.cwd(), 'tasks.csv');
const HEADERS = ['date', 'type', 'number', 'name', 'timeSpent', 'comments'];

function todayStr() {
  const d = new Date();
  return `${d.getMonth() + 1}/${d.getDate()}/${d.getFullYear()}`;
}

export function isValidDate(str) {
  if (!str) return false;
  const parts = str.split('/');
  if (parts.length !== 3) return false;
  const [m, d, y] = parts.map(Number);
  return m >= 1 && m <= 12 && d >= 1 && d <= 31 && y >= 2000;
}

function ensureDate(task) {
  if (!isValidDate(task.date)) {
    return { ...task, date: todayStr() };
  }
  return task;
}

function ensureFile() {
  if (!fs.existsSync(CSV_PATH)) {
    fs.writeFileSync(CSV_PATH, HEADERS.join(',') + '\n', 'utf8');
  }
}

export function loadTasks() {
  ensureFile();
  const content = fs.readFileSync(CSV_PATH, 'utf8');
  const result = Papa.parse(content, { header: true, skipEmptyLines: true });
  return result.data.map(row => {
    const cleaned = {};
    for (const key of HEADERS) {
      cleaned[key] = (row[key] || '').trim();
    }
    return cleaned;
  });
}

export function saveTasks(tasks) {
  const csv = Papa.unparse(tasks, { columns: HEADERS });
  fs.writeFileSync(CSV_PATH, csv + '\n', 'utf8');
}

export function addTask(task) {
  const tasks = loadTasks();
  tasks.push(ensureDate(task));
  saveTasks(tasks);
}

export function addTasks(newTasks) {
  const tasks = loadTasks();
  tasks.push(...newTasks.map(ensureDate));
  saveTasks(tasks);
}

export function updateTask(index, updated) {
  const tasks = loadTasks();
  if (index >= 0 && index < tasks.length) {
    tasks[index] = { ...tasks[index], ...updated };
    saveTasks(tasks);
  }
}

export function deleteTask(index) {
  const tasks = loadTasks();
  if (index >= 0 && index < tasks.length) {
    tasks.splice(index, 1);
    saveTasks(tasks);
  }
}

File: src/internal/store/store.go
package store

import (
	"encoding/csv"
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/aliaksandrZh/worklog/src/internal/model"
	"github.com/aliaksandrZh/worklog/src/internal/timeutil"
)

// TaskReader can load tasks.
type TaskReader interface {
	LoadTasks() ([]model.Task, error)
}

// TaskWriter can modify tasks.
type TaskWriter interface {
	AddTask(task model.Task) error
	AddTasks(tasks []model.Task) error
	UpdateTask(index int, updates map[string]string) error
	DeleteTask(index int) error
}

// TaskRepository combines reading and writing.
type TaskRepository interface {
	TaskReader
	TaskWriter
}

// CSVStore implements TaskRepository using a CSV file.
type CSVStore struct {
	Path string
}

// New creates a CSVStore at the default location (tasks.csv in cwd).
func New() *CSVStore {
	return &CSVStore{Path: filepath.Join(".", "tasks.csv")}
}

// NewWithPath creates a CSVStore at the given path.
func NewWithPath(path string) *CSVStore {
	return &CSVStore{Path: path}
}

// IsValidDate checks if a date string is valid M/D/YYYY.
func IsValidDate(str string) bool {
	if str == "" {
		return false
	}
	parts := strings.Split(str, "/")
	if len(parts) != 3 {
		return false
	}
	m, err1 := strconv.Atoi(parts[0])
	d, err2 := strconv.Atoi(parts[1])
	y, err3 := strconv.Atoi(parts[2])
	if err1 != nil || err2 != nil || err3 != nil {
		return false
	}
	return m >= 1 && m <= 12 && d >= 1 && d <= 31 && y >= 2000
}

func ensureDate(task model.Task) model.Task {
	if !IsValidDate(task.Date) {
		task.Date = timeutil.TodayStr()
	}
	return task
}

func (s *CSVStore) ensureFile() error {
	if _, err := os.Stat(s.Path); os.IsNotExist(err) {
		dir := filepath.Dir(s.Path)
		if dir != "." && dir != "" {
			if err := os.MkdirAll(dir, 0o755); err != nil {
				return err
			}
		}
		f, err := os.Create(s.Path)
		if err != nil {
			return err
		}
		w := csv.NewWriter(f)
		_ = w.Write(model.Headers)
		w.Flush()
		f.Close()
	}
	return nil
}

// LoadTasks reads all tasks from the CSV file.
func (s *CSVStore) LoadTasks() ([]model.Task, error) {
	if err := s.ensureFile(); err != nil {
		return nil, err
	}
	f, err := os.Open(s.Path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	r := csv.NewReader(f)
	records, err := r.ReadAll()
	if err != nil {
		return nil, err
	}

	if len(records) == 0 {
		return nil, nil
	}

	// Skip header
	header := records[0]
	colIndex := make(map[string]int)
	for i, h := range header {
		colIndex[strings.TrimSpace(h)] = i
	}

	var tasks []model.Task
	for _, row := range records[1:] {
		t := model.Task{}
		if idx, ok := colIndex["date"]; ok && idx < len(row) {
			t.Date = strings.TrimSpace(row[idx])
		}
		if idx, ok := colIndex["type"]; ok && idx < len(row) {
			t.Type = strings.TrimSpace(row[idx])
		}
		if idx, ok := colIndex["number"]; ok && idx < len(row) {
			t.Number = strings.TrimSpace(row[idx])
		}
		if idx, ok := colIndex["name"]; ok && idx < len(row) {
			t.Name = strings.TrimSpace(row[idx])
		}
		if idx, ok := colIndex["timeSpent"]; ok && idx < len(row) {
			t.TimeSpent = strings.TrimSpace(row[idx])
		}
		if idx, ok := colIndex["comments"]; ok && idx < len(row) {
			t.Comments = strings.TrimSpace(row[idx])
		}
		tasks = append(tasks, t)
	}
	return tasks, nil
}

// SaveTasks writes all tasks to the CSV file.
func (s *CSVStore) SaveTasks(tasks []model.Task) error {
	f, err := os.Create(s.Path)
	if err != nil {
		return err
	}
	defer f.Close()

	w := csv.NewWriter(f)
	_ = w.Write(model.Headers)
	for _, t := range tasks {
		_ = w.Write([]string{t.Date, t.Type, t.Number, t.Name, t.TimeSpent, t.Comments})
	}
	w.Flush()
	return w.Error()
}

// AddTask appends a single task (with date validation).
func (s *CSVStore) AddTask(task model.Task) error {
	tasks, err := s.LoadTasks()
	if err != nil {
		return err
	}
	tasks = append(tasks, ensureDate(task))
	return s.SaveTasks(tasks)
}

// AddTasks appends multiple tasks (with date validation).
func (s *CSVStore) AddTasks(newTasks []model.Task) error {
	tasks, err := s.LoadTasks()
	if err != nil {
		return err
	}
	for _, t := range newTasks {
		tasks = append(tasks, ensureDate(t))
	}
	return s.SaveTasks(tasks)
}

// UpdateTask updates fields of the task at the given index.
func (s *CSVStore) UpdateTask(index int, updates map[string]string) error {
	tasks, err := s.LoadTasks()
	if err != nil {
		return err
	}
	if index < 0 || index >= len(tasks) {
		return fmt.Errorf("index %d out of range", index)
	}
	t := tasks[index]
	for k, v := range updates {
		switch k {
		case "date":
			t.Date = v
		case "type":
			t.Type = v
		case "number":
			t.Number = v
		case "name":
			t.Name = v
		case "timeSpent":
			t.TimeSpent = v
		case "comments":
			t.Comments = v
		}
	}
	tasks[index] = t
	return s.SaveTasks(tasks)
}

// DeleteTask removes the task at the given index.
func (s *CSVStore) DeleteTask(index int) error {
	tasks, err := s.LoadTasks()
	if err != nil {
		return err
	}
	if index < 0 || index >= len(tasks) {
		return fmt.Errorf("index %d out of range", index)
	}
	tasks = append(tasks[:index], tasks[index+1:]...)
	return s.SaveTasks(tasks)
}
